cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
 	wg.Wait()
 	return resultingKubeConfig
 }
+
+// certChain is an array of *x509.Certificate with a helper to plug into the tls VerifyPeerCertificate hook
+type certChain []*x509.Certificate
+
+// saveCertificateChain parses and saves the array of DER-encoded certificates into a certChain
+func (c *certChain) saveCertificateChain(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+	// Make sure this wrapper is being used correctly
+	if len(*c) != 0 {
+		return errInsecureClientUsedMoreThanOnce
+	}
+
+	// Do a quick sanity check (this should never happen)
+	if len(rawCerts) < 1 {
+		return errInsecureClientEmptyCertChain
+	}
+
+	// Parse and collect each certificate
+	for _, rawCert := range rawCerts {
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			return err
+		}
+		*c = append(*c, cert)
+	}
+
+	return nil
+}
+
+// withInsecureHTTPClient creates a temporary http.Client, passes it to doRequest(...),
+// which should perform a single HTTP request and return the response. The temporary
+// client will not validate TLS certificates in real time, but instead will capture
+// and return the certificates for post-hoc validation.
+func withInsecureHTTPClient(
+	doRequest func(*http.Client) (*http.Response, error),
+) (*http.Response, []*x509.Certificate, error) {
+
+	// certificates will collect the certificate chain presented by the server
+	var certificates certChain
+
+	// Create an HTTP client with a custom TLS transport that saves off the certificate chain for later validation
+	insecureClient := http.Client{
```

Comment:
That makes sense. I think it shouldn't be a huge issue for `kubeadm` since it's relatively short-lived. That's still pretty non-ideal, though.

I'll see if I can find a way to rework this so the http.Client is long lived.